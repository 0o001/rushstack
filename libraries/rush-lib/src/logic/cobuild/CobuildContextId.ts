// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { RushConstants } from '../RushConstants';

export interface IGenerateCobuildContextIdOptions {
  environment: NodeJS.ProcessEnv;
}

/**
 * Calculates the cache entry id string for an operation.
 * @beta
 */
export type GetCobuildContextIdFunction = (options: IGenerateCobuildContextIdOptions) => string;

export class CobuildContextId {
  private constructor() {}

  public static parsePattern(pattern?: string): GetCobuildContextIdFunction {
    if (!pattern) {
      return () => '';
    } else {
      const resolvedPattern: string = pattern.trim();

      return (options: IGenerateCobuildContextIdOptions) => {
        const { environment } = options;
        return this._expandWithEnvironmentVariables(resolvedPattern, environment);
      };
    }
  }

  private static _expandWithEnvironmentVariables(pattern: string, environment: NodeJS.ProcessEnv): string {
    const missingEnvironmentVariables: Set<string> = new Set<string>();
    const expandedPattern: string = pattern.replace(
      /\$\{([^\}]+)\}/g,
      (match: string, variableName: string): string => {
        const variable: string | undefined =
          variableName in environment ? environment[variableName] : undefined;
        if (variable !== undefined) {
          return variable;
        } else {
          missingEnvironmentVariables.add(variableName);
          return match;
        }
      }
    );
    if (missingEnvironmentVariables.size) {
      throw new Error(
        `The "cobuildContextIdPattern" value in ${
          RushConstants.cobuildFilename
        } contains missing environment variable${
          missingEnvironmentVariables.size > 1 ? 's' : ''
        }: ${Array.from(missingEnvironmentVariables).join(', ')}`
      );
    }

    return expandedPattern;
  }
}
