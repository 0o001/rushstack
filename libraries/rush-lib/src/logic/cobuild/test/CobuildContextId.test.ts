// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { CobuildContextId } from '../CobuildContextId';

describe(CobuildContextId.name, () => {
  describe('Valid pattern names', () => {
    it('expands a environment variable', () => {
      const contextId: string = CobuildContextId.parsePattern('context-${MR_ID}-${AUTHOR_NAME}')({
        environment: {
          MR_ID: '123',
          AUTHOR_NAME: 'Mr.example'
        }
      });
      expect(contextId).toEqual('context-123-Mr.example');
    });
  });

  describe('Invalid pattern names', () => {
    it('throws an error if a environment variable is missing', () => {
      expect(() =>
        CobuildContextId.parsePattern('context-${MR_ID}-${AUTHOR_NAME}')({
          environment: {
            MR_ID: '123'
          }
        })
      ).toThrowErrorMatchingSnapshot();
    });
    it('throws an error if multiple environment variables are missing', () => {
      expect(() =>
        CobuildContextId.parsePattern('context-${MR_ID}-${AUTHOR_NAME}')({
          environment: {}
        })
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
