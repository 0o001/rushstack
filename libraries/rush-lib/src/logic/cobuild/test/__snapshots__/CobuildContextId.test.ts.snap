// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CobuildContextId Invalid pattern names throws an error if a environment variable is missing 1`] = `"The \\"cobuildContextIdPattern\\" value in cobuild.json contains missing environment variable: AUTHOR_NAME"`;

exports[`CobuildContextId Invalid pattern names throws an error if multiple environment variables are missing 1`] = `"The \\"cobuildContextIdPattern\\" value in cobuild.json contains missing environment variables: MR_ID, AUTHOR_NAME"`;
